feat(hardware): add a update routine and script for rear-panel

ryanthecoder opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview
Tested on the new rear-panel, script takes the following steps given a .bin file from the firmware build

put the device in dfu mode
disconnect the serial port
stop the messenger
use dfu-util to find and update the firmware image
reconnect to the serial port
restart the messenger

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

# Test Plan

<!--
Use this section to describe the steps that you took to test your Pull Request.
If you did not perform any testing provide justification why.

OT-3 Developers: You should default to testing on actual physical hardware.
Once again, if you did not perform testing against hardware, justify why.

Note: It can be helpful to write a test plan before doing development

Example Test Plan (HTTP API Change)

- Verified that new optional argument `dance-party` causes the robot to flash its lights, move the pipettes,
then home.
- Verified that when you omit the `dance-party` option the robot homes normally
- Added protocol that uses `dance-party` argument to G-Code Testing Suite
- Ran protocol that did not use `dance-party` argument and everything was successful
- Added unit tests to validate that changes to pydantic model are correct

-->

# Changelog

<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

ryanthecoder committed:
add an update routine for the rear panel to the firmware update code

ryanthecoder committed:
add a fw update script for rear panel

ryanthecoder committed:
breakout serial driver building

ryanthecoder committed:
python on the bot cant find this function unless i import the whole path

ryanthecoder committed:
set task to none so messenger can be stoped then started again

ryanthecoder committed:
fix the build for device info response to account for optional revision

ryanthecoder committed:
fix the update script, now fully tested

ryanthecoder committed:
ignore untyped import

ryanthecoder committed:
remove unneeded long call

ryanthecoder committed:
few linting things

ryanthecoder committed:
for the int base use ints instead of bytes to build

ryanthecoder committed:
filter the listener to just device info responses

ryanthecoder committed:
fix and add test for messenger start and stop

sfoster1 commented on hardware/opentrons_hardware/firmware_update/run.py:
```diff
@@ -168,3 +175,199 @@ async def run_updates(
+            if pid in line:
+                logger.info(f"Found device with PID {pid}")
+                devices_found += 1
+                serial = line[(line.find("serial=") + 7) :]
```
can we do all the filtering by serial? do we know how the serial is established?

sfoster1 commented on hardware/opentrons_hardware/firmware_update/run.py:
```diff
@@ -168,3 +175,199 @@ async def run_updates(
+
+    Unlike other firmware upload methods, this one doesn't take a `port` argument since
+    the module isn't recognized as a cdc device in dfu mode and hence doesn't get
+    a port. The firmware upload utility, dfu-util, looks for the specific module
```
remove the "module" references

sfoster1 commented on hardware/opentrons_hardware/firmware_update/run.py:
```diff
@@ -168,3 +175,199 @@ async def run_updates(
+    async def _request_version(self) -> None:
+        await self._messenger.send(DeviceInfoRequest())
+        # give time for the device to respond
+        await asyncio.sleep(1)
```
We should not do this - we should wait for the message to come back. We can use an asyncio.event or asyncio.queue for this.

sfoster1 commented on hardware/opentrons_hardware/firmware_update/run.py:
```diff
@@ -168,3 +175,199 @@ async def run_updates(
+        # give time for the device to respond
+        await asyncio.sleep(1)
+
+    async def run_update(self) -> bool:
```
This process should look a lot more like the process for canbus connected devices, since it will run in similar ways. I would argue, in fact, that because this process will happen at the same time as all the other updates it should be just another method of the existing updater class, that takes a status queue and emits progress updates just as a canbus updater does.

sfoster1 commented on hardware/opentrons_hardware/scripts/update_rpfw.py:
```diff
@@ -0,0 +1,102 @@
+"""Firmware update script for the rear panel."""
```
Similarly, let's make this handled by the already-existing update script.

sfoster1 reviewed (changes requested):
I think we need to integrate this a little more tightly into the already-existing update machinery. While we don't have to go all the way to making it perfect in this PR, adding infrastructure like this means it'll be deleted in the next PR, so let's add it as new internal functions to existing classes.

ryanthecoder committed:
remove references to modules

ryanthecoder committed:
use events to wait for version resposne

ryanthecoder committed:
make the rear panel firmware update emit progress like the can updates

ryanthecoder committed:
integrate the rearpanel with the existing update fw script

ryanthecoder committed:
combine the usb and can updating into one class

ryanthecoder committed:
fix the old can update tests

ryanthecoder committed:
fix an issue where this test intermentaly fails due to the race condition

sfoster1 commented on hardware/opentrons_hardware/firmware_update/run.py:
```diff
@@ -45,20 +155,110 @@ def __init__(
-    async def _run_update(
+    def __call__(self, message: BinaryMessageDefinition) -> None:
+        """Function called when bus receives a device info response."""
+        logger.info(f"received version info from device {message}")
```
if this doesn't do anything, is there a reason to have it?

ryanthecoder replied:
I can take it out, this was how I was making sure that the flashing was successful and the reporting a new version

sfoster1 reviewed (approved):
Generally looking good if tested and works!

ryanthecoder committed:
refactor the way RunUpdates works with the ot3controller and make it work more seamlessly for now
there will be further improvements in a further pr that include:
getting device info for rearpanel on startup
checking if the rear panel fw is up to date
any typing changes that need to happen to make the rear-panel target play nice with the node_id targets

ryanthecoder committed:
remove listener I used to debug rear-panel updates

ryanthecoder committed:
update the progress tracker for usb updates and put the messenger.stop in the right spot

ryanthecoder committed:
fix one error that was mine and one that wasn't in update_fws

ryanthecoder committed:
format/lint

ryanthecoder commented:
Fully tested now with both the update_fw and update_fws scripts
